gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/proc/KillDagProc.java

```diff
@@ -0,0 +1,125 @@
+import java.util.concurrent.Future;
+
+import org.apache.directory.api.util.Strings;
+
+import com.google.common.collect.Maps;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.gobblin.configuration.ConfigurationKeys;
+import org.apache.gobblin.metrics.event.TimingEvent;
+import org.apache.gobblin.service.modules.flowgraph.Dag;
+import org.apache.gobblin.service.modules.flowgraph.DagNodeId;
+import org.apache.gobblin.service.modules.orchestration.DagManagementStateStore;
+import org.apache.gobblin.service.modules.orchestration.DagManagerUtils;
+import org.apache.gobblin.service.modules.orchestration.TimingEventUtils;
+import org.apache.gobblin.service.modules.orchestration.task.KillDagTask;
+import org.apache.gobblin.service.modules.spec.JobExecutionPlan;
+
+import static org.apache.gobblin.service.ExecutionStatus.CANCELLED;
+
+
+/**
+ * An implementation for {@link DagProc} that kills all the nodes of a dag.
+ * If the dag action has job name set, then it kills only that particular job/dagNode.
+ */
+@Slf4j
+public class KillDagProc extends DagProc<Optional<Dag<JobExecutionPlan>>> {
+  private final Optional<DagNodeId> dagNodeId;
+
+  public KillDagProc(KillDagTask killDagTask) {
+    super(killDagTask);
+    this.dagNodeId = Strings.isEmpty(this.dagTask.getDagAction().getJobName())
+        ? Optional.empty()
+        : Optional.of(new DagNodeId(this.dagTask.getDagAction().getFlowGroup(), this.dagTask.getDagAction().getFlowName(),
+        Long.parseLong(this.dagTask.getDagAction().getFlowExecutionId()),
+        this.dagTask.getDagAction().getFlowGroup(), this.dagTask.getDagAction().getJobName()));
+  }
+
+  @Override
+  protected Optional<Dag<JobExecutionPlan>> initialize(DagManagementStateStore dagManagementStateStore)
+      throws IOException {
+   return dagManagementStateStore.getDag(getDagId());
+  }
+
+  @Override
+  protected void act(DagManagementStateStore dagManagementStateStore, Optional<Dag<JobExecutionPlan>> dag)
+      throws IOException {
+    if (!dag.isPresent()) {
+      log.error("Did not find Dag with id {}, it might be already cancelled/finished and thus cleaned up from the store.", getDagId());
+      return;
+    }
+
+    dag.get().setFlowEvent(TimingEvent.FlowTimings.FLOW_CANCELLED);
+    dag.get().setMessage("Flow killed by request");
+
+    dagManagementStateStore.checkpointDag(dag.get());
+
+    if (this.dagNodeId.isPresent()) {
+      cancelDagNode(dagManagementStateStore.getDagNodeWithJobStatus(this.dagNodeId.get()).getLeft().get());
+      return;
+    }
+
+    List<Dag.DagNode<JobExecutionPlan>> dagNodesToCancel = dag.get().getNodes();
+    log.info("Found {} DagNodes to cancel (DagId {}).", dagNodesToCancel.size(), getDagId());
+
+    for (Dag.DagNode<JobExecutionPlan> dagNodeToCancel : dagNodesToCancel) {
+      cancelDagNode(dagNodeToCancel);
+      // todo - why was it not being cleaned up in DagManager?
+      dagManagementStateStore.deleteDagNodeState(getDagId(), dagNodeToCancel);
```

Comment:
are you saying the current impl is not calling the equivalent of `deleteDagNodeState`?

Reply:
yea, it's like // todo - (understand) why was it not being cleaned up in the current impl.
we can delete the comment, it maybe a bug in current impl, it is not too problemetic  one because it is a memory based impl anyway.